### ISSUE TITLE
Call to /rest/system/error breaks when syncthing is reverse-proxied in a sub-folder

### ISSUE THREAD

**Issue by javmorin** [syncthing/syncthing#4160]
### Version Information
Syncthing Version: v0.14.28
OS Version: Arch Linux
Browser Version: Chrome (59.0.3071.47)

This has been occurring for quite some time, but I have only now gotten around to reporting it. I've set up an nginx reverse proxy (with https+auth) on my server with syncthing being "placed" in a sub-folder. If I open the GUI and leave it up, it reports a connection failure as soon as the first polling event occurs. Looking at the debug console, I see that it is the call to /rest/system/error that is returning a 404, all others that should be running fine, are. Looking at the request, it is apparent that it is missing the prefix path/base path where the syncthing is "placed" (it's using /rest/system/error as apposed to /syncthing/rest/system/error). 

I added a rewrite in my nginx config to handle the error path without the folder/prefix, reloaded the server and browser, and it works just fine; no connection error dialog.

I suspect this is a 20 second fix, but not being a JS person, I am not sure where to look to fix it.

**Comment by AudriusButkevicius** [comment]
This is no way different from any other call in the API:

https://github.com/syncthing/syncthing/blob/ac510b26e259271d2d0b530a7d3e46a3169a8cf1/gui/default/syncthing/core/syncthingController.js#L512

and urlbase is always relative, never absolute or prefixed with a slash:

https://github.com/syncthing/syncthing/blob/161326c5489d000972a6846564f0ce12779bd8f2/gui/default/syncthing/app.js#L18

**Comment by AudriusButkevicius** [comment]
Did you follow this:

https://docs.syncthing.net/users/reverseproxy.html#nginx

Or did you invent your own nginx config?

**Comment by javmorin** [comment]
I started with that config, yes. I had to add a couple rewrite commands in because syncthing kept redirecting asset requests to the root. I have just tried reverting to the suggested config again, and it still does this. Might be an issue with some of the other proxy settings that I have, but I am not certain. The config I have does work other than this error call so.

I am sure it is just the error call that is failing. I can see the others are returning properly and from the expected (prefixed) path. The only other failing call is upgrade, but I understand that is normal.

I will do a bit more digging.

**Comment by calmh** [comment]
We take a fair amount of care to not ever do absolute requests, so that this thing should work out of the box. And it's been that way for quite a while. So I would be very surprised if there was anything that actually did requests towards the root if the actual gui is being served from a subpath.

**Comment by javmorin** [comment]
BTW, speaking to the reverse proxy config, I have to add the following in order to make it work properly.
```
location /syncthing {
    proxy_pass              http://localhost:8384/;

    # make sure the syncthing path ends in a /
    rewrite ^/syncthing$ /syncthing/ permanent;

    # force syncthing into thinking it is the root to avoid asset redirects
    rewrite /syncthing/(.*) /$1 break;

    proxy_read_timeout      600s;
    proxy_send_timeout      600s;
}
```

**Comment by calmh** [comment]
That's a weird config and the rewrite should do nothing but confuse things. Try our example configs, they work fine for a lot of people.

**Comment by javmorin** [comment]
Well don't I feel sheepish now. I decided to try the UI in Firefox (which I normally don't use), and it worked right from the start. Correct URLs and everything. I thought that was odd, went back to Chrome with a fresh session and it was still using the wrong URL. I tried doing a forced reload, still the wrong URL. Went in and disabled the cache in dev tools, and now it works! 

As mentioned before, this has been going on for *months* on multiple machines. Because of my troubles with the nginx config it must have cached in the machines before I "fixed" the config. No idea why/how Chrome was caching the wrong location of just the error page for so long on multiple machines though. It's too smart for it's own good I suppose.

Sorry for the red-herring issue. Closing it now.

**Comment by javmorin** [comment]
@calmh 
I just tried it again, and to make sure, I tried making the requests with curl to avoid any caching issues. No dice. If others aren't having issues, that's good, no need to update the docs. Not sure why my nginx config is snowflake, but that rewrite is required for syncthing to work properly in a sub-folder on my server.

**Comment by canton7** [comment]
Permanent rewrites are pretty much permanent. It's really, really hard to get browsers to forget them. I'd advise never returning 301 or 308 unless you've debugged your configuration, and you're absolutely sure it will never ever change. Go with 307 instead.

**Comment by javmorin** [comment]
@AudriusButkevicius, @calmh Found the source of the troubles in my nginx config, and now the config from the wiki works just fine. 

@canton7 While I appreciate the insight, if you review my config again, you will note that the permanent redirect (in fact the only redirect) is for adding a trailing slash if I use /syncthing instead of /syncthing/. The other rewrite only changes the representation of the request path that is forwarded over the proxy to the syncthing daemon.

**Comment by canton7** [comment]
> @AudriusButkevicius, @calmh Found the source of the troubles in my nginx config, and now the config from the docs works just fine.

What was it, out of curiosity (and for anyone else hitting the same issue)?

**Comment by javmorin** [comment]
>
> What was it, out of curiosity (and for anyone else hitting the same issue)?
>
>
I should first point out this had nothing to do with the issue I reported,
that was strictly a caching problem.

The problem I had with the config from the docs was due to the fact that I
didn't have a slash at the end of the location block. I had "/syncthing"
instead of "/syncthing/" because I had my redirect within the block. nginx
behavior has a bunch of little quirks like that. I've been using nginx in
production at work and at home for > 10 years but sometimes those quirks
catch me. I didn't worry about it because the config I had worked just fine.
